### PR TITLE
Task timing with FreeRTOS

### PR DESCRIPTION
**Pull request by upadhyaydhruv**
FreeRTOS vTaskDelayUntil works with blinky representing Telemetry, AttitudeManager, and PathManager. Still need to get it done for sensorfusion but the concurrency for the first three has been tested on the breakout board

### CODE REVIEW

**Review by sahil-kale — CHANGES_REQUESTED** [review]
Can we uncommit the files in .cache/?

Also, build is failing

File: Autopilot/AttitudeManager/Inc/attitudeStateClasses.hpp
Comment (Fberbari) [inline Autopilot/AttitudeManager/Inc/attitudeStateClasses.hpp]: Build fails because of this stuff, I think. Don't comment this state. Rather just create a dummy implementation for sensor fusion_getResult that does nothing. Commenting out states fucks with the state control logic and breaks the unit tests.

File: Autopilot/AttitudeManager/Src/SendInstructionsToSafety.cpp
Comment (Fberbari) [inline Autopilot/AttitudeManager/Src/SendInstructionsToSafety.cpp]: use <cstdint> rather than the deprecated c equivalent.

File: Autopilot/AttitudeManager/Src/attitudeStateClasses.cpp
Comment (Fberbari) [inline Autopilot/AttitudeManager/Src/attitudeStateClasses.cpp]: same sort of thing. We can't merge changes that comment out previous changes like this, It breaks the flow. Gotta create a dummy implementation for them rather than do this.

File: Autopilot/Inc/SensorFusion.hpp
Comment (Fberbari) [inline Autopilot/Inc/SensorFusion.hpp]: it is no longer necessary to drag the struct keyword around in c++ 11. Why did you put them in. Was something failling ?

File: Autopilot/PathManager/Src/PathManagerInterface.cpp
Comment (Fberbari) [inline Autopilot/PathManager/Src/PathManagerInterface.cpp]: there's a caveat here. Calling execute once only executes a single state. But you want to run the full state machine once per FREERTOS thread execution. There's a mechanism to know wether the state machine is done executing or not. Take a look at how it's called in ZeroPilot-SW\Autopilot\Simulation\SimDriver\SimMain.cpp

File: Autopilot/Src/SensorFusion.cpp
Comment (Fberbari) [inline Autopilot/Src/SensorFusion.cpp:98]: Oh, you actually did create this. So you shouldn't have to comment out calls to sensor fusion in Path manager, right ?

File: Autopilot/boardfiles/Src/freertos.c
Comment (Fberbari) [inline Autopilot/boardfiles/Src/freertos.c]: I think this is you issue actually. Your'e including SensorFusion.hpp here. You shouldn't have to. you only need the interface.

File: Autopilot/boardfiles/Src/freertos.c
Comment (Fberbari) [inline Autopilot/boardfiles/Src/freertos.c]: interfaces should all be .h, no ? Whole point is to have them work with C.

File: Autopilot/boardfiles/Src/freertos.c
Comment (Fberbari) [inline Autopilot/boardfiles/Src/freertos.c]: niiiice. Quick tip. "TODO" is sortof the universal comment for "I'll have to figure this out later", YOu should use it. That way, in the future, we can just search the project for TODOs and see what little things need to be addressed.

File: Autopilot/boardfiles/Src/freertos.c
Comment (Fberbari) [inline Autopilot/boardfiles/Src/freertos.c]: good stuff. Just a question, "osPriorityBelowNorma" what does that mean and why did you chose it ?

File: Autopilot/boardfiles/Src/freertos.c
Comment (Fberbari) [inline Autopilot/boardfiles/Src/freertos.c:210]: Nice, feel free to toss in whatever blinky lights you like :)

File: Autopilot/boardfiles/Src/main.c
Comment (Fberbari) [inline Autopilot/boardfiles/Src/main.c]: If you wanna get rd of the watchdog, gotta do it from the IOC file. Modifying code inside the generated section isn't a good ifdea since it would get written over every time.

**Review by Fberbari — CHANGES_REQUESTED** [review]
Good work! The bulk of it looks good. A few things to address. Also, looks like the cache files are still in the commit.

**Comment by upadhyaydhruv** [comment]
Thanks for the review, I'll try and get the changes fixed for Thursday

File: Autopilot/AttitudeManager/Src/SendInstructionsToSafety.cpp
Comment (upadhyaydhruv) [reply Autopilot/AttitudeManager/Src/SendInstructionsToSafety.cpp]: would that be cstdint? (for some reason github gets rid of the text between <>)

File: Autopilot/Inc/SensorFusion.hpp
Comment (upadhyaydhruv) [reply Autopilot/Inc/SensorFusion.hpp]: I think since the functions in sensorfusion are called by freertos they expect it to be in C-Style, because it was failing without the struct keyword

File: Autopilot/boardfiles/Src/freertos.c
Comment (upadhyaydhruv) [reply Autopilot/boardfiles/Src/freertos.c]: tbh this was a pretty arbitrary decision, there are different priority levels for the threads, where osPriorityRealtime is the highest and osPriorityBelowNormal is the lowest. I guess it depends on how important telemetry is compared to attitudemanager, sensorfusion, and pathmanager which would also influence the priorities of those ones. Would you have any suggestions for the priority?

File: Autopilot/Src/SensorFusion.cpp
Comment (upadhyaydhruv) [reply Autopilot/Src/SensorFusion.cpp:98]: yeah true, I think I just forgot to uncomment those earlier parts haha

File: Autopilot/PathManager/Src/PathManagerInterface.cpp
Comment (upadhyaydhruv) [reply Autopilot/PathManager/Src/PathManagerInterface.cpp]: So looking at the solution in simulation that you guided me too, it seems as if they use a while loop. Is there a possible non-blocking solution to this or is this good enough for our purposes?

File: Autopilot/AttitudeManager/Inc/attitudeManagerInterface.h
Comment (Pop0097) [inline Autopilot/AttitudeManager/Inc/attitudeManagerInterface.h]: newline needed

File: Autopilot/AttitudeManager/Src/attitudeManagerInterface.cpp
Comment (Pop0097) [inline Autopilot/AttitudeManager/Src/attitudeManagerInterface.cpp]: Please fix the brackets:

```cpp
do {

} while ();
```

File: Autopilot/Inc/Interchip_A.h
Comment (Pop0097) [inline Autopilot/Inc/Interchip_A.h]: newline

File: Autopilot/PathManager/Inc/PathManagerInterface.h
Comment (Pop0097) [inline Autopilot/PathManager/Inc/PathManagerInterface.h]: go through all the files and fix all the newline warnings pwease

File: Autopilot/PathManager/Src/PathManagerInterface.cpp
Comment (Pop0097) [inline Autopilot/PathManager/Src/PathManagerInterface.cpp]:
```cpp
do {

} while ();
```

**Review by Pop0097 — COMMENTED** [review]
Nice work so far!

**Comment by Pop0097** [comment]
We will merge once testing is complete

File: Autopilot/Inc/SensorFusion.hpp
Comment (Fberbari) [reply Autopilot/Inc/SensorFusion.hpp]: hm, I see. This should be taken care of as part of the interface between our functions and FREERTOS. It needs to be, otherwise you'l lbe dissallowing Tony from using C++.

File: Autopilot/PathManager/Src/PathManagerInterface.cpp
Comment (Fberbari) [reply Autopilot/PathManager/Src/PathManagerInterface.cpp]: Not sure what you mean by blocking/non blocking here. Usually, blocking means a thread busy waits without doing any meaningful work. THat's not what happens here. All the while loop ensures is that the state machine is called the appropriate number of times to execute 1 full cycle. (It should execute 1 full cycle per FREERTOS thread call)

File: Autopilot/boardfiles/Src/freertos.c
Comment (Fberbari) [reply Autopilot/boardfiles/Src/freertos.c]: Ah I see. If we do things right, threads should never step on each other's toes. In fact, we should make sure that each thread has finished a single execution before FREERTOS moves on to the next one. That will likely be done via a mutex. But we should talk about that later. ATM, having them all normal is fine

File: Autopilot/AttitudeManager/Src/attitudeManager.cpp
Comment (Pop0097) [inline Autopilot/AttitudeManager/Src/attitudeManager.cpp]: Will have to remember to change this back

File: Autopilot/PathManager/Src/PathManagerInterface.cpp
Comment (Pop0097) [inline Autopilot/PathManager/Src/PathManagerInterface.cpp]: newline stuff lol.

File: Autopilot/PathManager/Src/PathManagerInterface.cpp
Comment (Pop0097) [inline Autopilot/PathManager/Src/PathManagerInterface.cpp]: only need one return statement

File: Autopilot/SensorFusion/Inc/sensorFusionInterface.hpp
Comment (Pop0097) [inline Autopilot/SensorFusion/Inc/sensorFusionInterface.hpp]: For these header names, I would rather you separate words with underscores. So this becomes `SENSOR_FUSION_INTERFACE_H`. It just makes everything easier to read :))

File: Autopilot/SensorFusion/Inc/sensorFusionInterface.h
Comment (Pop0097) [inline Autopilot/SensorFusion/Inc/sensorFusionInterface.h]: Duplicate file

Comment (Pop0097) [reply Autopilot/SensorFusion/Inc/sensorFusionInterface.h]: I say we should stick with sensorFusionInterface.hpp

File: Autopilot/boardfiles/Src/freertos.c
Comment (Pop0097) [inline Autopilot/boardfiles/Src/freertos.c:272]: Is there an error case we want to consider?

File: Autopilot/boardfiles/Src/freertos.c
Comment (Pop0097) [inline Autopilot/boardfiles/Src/freertos.c]: any chance we can name this telemetryManagerExecute for consistency

**Review by Pop0097 — COMMENTED** [review]
Looking good! How are hardware tests going?

File: Autopilot/boardfiles/Src/freertos.c
Comment (upadhyaydhruv) [reply Autopilot/boardfiles/Src/freertos.c]: Will do

File: Autopilot/boardfiles/Src/freertos.c
Comment (upadhyaydhruv) [reply Autopilot/boardfiles/Src/freertos.c:272]: I'll look into it, I think if there is an error I'll just set catastrophicError to true similar to the other threads

File: Autopilot/SensorFusion/Inc/sensorFusionInterface.h
Comment (upadhyaydhruv) [reply Autopilot/SensorFusion/Inc/sensorFusionInterface.h]: Sounds good, I think I forgot to delete this one before committing

File: Autopilot/SensorFusion/Inc/sensorFusionInterface.hpp
Comment (upadhyaydhruv) [reply Autopilot/SensorFusion/Inc/sensorFusionInterface.hpp]: sounds good, will do

File: Autopilot/PathManager/Src/PathManagerInterface.cpp
Comment (upadhyaydhruv) [reply Autopilot/PathManager/Src/PathManagerInterface.cpp]: ye I forgot this one too mb

**Comment by upadhyaydhruv** [comment]
> Looking good! How are hardware tests going?

Testing it, it seemed to work well and as expected. All the blinkies representing the threads worked, and changing all the states to FAILURE_MODE did stop running the Interchip_Run() loop.

File: Autopilot/boardfiles/Src/freertos.c
Comment (Pop0097) [inline Autopilot/boardfiles/Src/freertos.c:6]: yeah, feel free to add your name wherever it is required :))

**Review by Pop0097 — APPROVED** [review]
This does look good to me, and the tests did pass. @Fberbari @sahil-kale can you take a look and remove your "Requested changes" at your earliest convenience?

**Review by Pop0097 — CHANGES_REQUESTED** [review]
Sorry, if you take a look at Travis you'll see the tests are not building. We will have to wait until that is fixed before merging. I should have a fix in very shortly, so once that is done I'll reapprove this

**Review by Pop0097 — APPROVED** [review]

**Review by sahil-kale — APPROVED** [review]

File: Autopilot/AttitudeManager/Inc/attitudeStateClasses.hpp
Comment (Fberbari) [inline Autopilot/AttitudeManager/Inc/attitudeStateClasses.hpp]: looks like no changes were actually made to this file. It shouldn't be in this review.

File: Autopilot/AttitudeManager/Src/attitudeManagerInterface.cpp
Comment (Fberbari) [inline Autopilot/AttitudeManager/Src/attitudeManagerInterface.cpp:18]: LMAO, your'e quite polite. I would have written "shit's gone to hell" XD

File: Autopilot/AttitudeManager/Src/attitudeStateClasses.cpp
Comment (Fberbari) [inline Autopilot/AttitudeManager/Src/attitudeStateClasses.cpp]: No changes here, you should take this file out of the review

File: Autopilot/CMakeLists.txt
Comment (Fberbari) [inline Autopilot/CMakeLists.txt:233]: This file, is this something you changed ? Or did you just forget to rebase ?

**Review by Fberbari — APPROVED** [review]
Approving, but pls address my comments!

File: Autopilot/CMakeLists.txt
Comment (upadhyaydhruv) [reply Autopilot/CMakeLists.txt:233]: I think I forgot to rebase for this one

**Pull request merged**
